vms/evm/warp/backend.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// Add stores a warp message in the database and cache.
+func (d *DB) Add(unsignedMsg *warp.UnsignedMessage) error {
+	msgID := unsignedMsg.ID()
+	log.Debug("Adding warp message to backend", "messageID", msgID)
+
+	// In the case when a node restarts, and possibly changes its bls key, the cache gets emptied but the database does not.
+	// So to avoid having incorrect signatures saved in the database after a bls key change, we save the full message in the database.
+	// Whereas for the cache, after the node restart, the cache would be emptied so we can directly save the signatures.
+	if err := d.db.Put(msgID[:], unsignedMsg.Bytes()); err != nil {
+		return fmt.Errorf("failed to put warp message in db: %w", err)
+	}
+
+	return nil
+}
+
+// Get retrieves a warp message from cache, offchain messages, or database.
+func (d *DB) Get(msgID ids.ID) (*warp.UnsignedMessage, error) {
+	if msg, ok := d.messageCache.Get(msgID); ok {
+		return msg, nil
+	}
+	if msg, ok := d.offchainAddressedCallMsgs[msgID]; ok {
+		return msg, nil
+	}
+
+	unsignedMessageBytes, err := d.db.Get(msgID[:])
+	if err != nil {
+		return nil, err
+	}
+
+	unsignedMessage, err := warp.ParseUnsignedMessage(unsignedMessageBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse unsigned message %s: %w", msgID.String(), err)
+	}
+	d.messageCache.Put(msgID, unsignedMessage)
+
+	return unsignedMessage, nil
+}
+
+// Signer signs warp messages and caches the signatures.
+type Signer struct {
+	warpSigner     warp.Signer
+	verifier       acp118.Verifier
+	signatureCache cache.Cacher[ids.ID, []byte]
+}
+
+// NewSigner creates a new warp message signer.
+func NewSigner(
+	warpSigner warp.Signer,
+	verifier acp118.Verifier,
+	signatureCache cache.Cacher[ids.ID, []byte],
+) *Signer {
+	return &Signer{
+		warpSigner:     warpSigner,
+		verifier:       verifier,
+		signatureCache: signatureCache,
+	}
+}
+
+// Sign verifies the warp message, signs it, and caches the signature.
+func (s *Signer) Sign(ctx context.Context, msg *warp.UnsignedMessage) ([]byte, error) {
+	// Check cache first
+	msgID := msg.ID()
+	if sig, ok := s.signatureCache.Get(msgID); ok {
+		return sig, nil
+	}
+
+	if err := s.verifier.Verify(ctx, msg, nil); err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrVerifyWarpMessage, err)
+	}
+
+	sig, err := s.warpSigner.Sign(msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign warp message: %w", err)
+	}
+
+	s.signatureCache.Put(msgID, sig)
+	return sig, nil
+}
+
+// Handler implements p2p.Handler and handles warp signature requests.
+// It hides the acp118.Verifier implementation as an implementation detail.
+type Handler struct {
+	*acp118.Handler
+}
+
+// NewHandler creates a new p2p warp signature request handler.
+func NewHandler(
+	signatureCache cache.Cacher[ids.ID, []byte],
+	verifier acp118.Verifier,
+	signer warp.Signer,
+) p2p.Handler {
+	return &Handler{
+		Handler: acp118.NewCachedHandler(signatureCache, verifier, signer),
+	}
+}
```

Comment:
It seems like we don't need this `Handler` type w/ the current api... so we should be able to get rid of this.